Can't bind functions taking char& as an argument

Dear Pybind11 developers, 
Looks like both latest `master` and `stable` seems to produce an error when trying to bind function taking `char &`.  Could you please looks this up?

Thanks,

```C++
#include <pybind11/pybind11.h>

void foo_char(char &) {}
void foo_bool(bool &) {}
void foo_int(int &) {}
void foo_long(long &) {}
void foo_long_long(long long &) {}
void foo_float(float &) {}
void foo_double(double &) {}

PYBIND11_PLUGIN(example) {
    pybind11::module m("example", "pybind11 example plugin");

	m.def("foo_char",      &foo_char);
	m.def("foo_bool",      &foo_bool);
	m.def("foo_int",       &foo_int);
	m.def("foo_long",      &foo_long);
	m.def("foo_long_long", &foo_long_long);
	m.def("foo_float",     &foo_float);
	m.def("foo_double",    &foo_double);

    return m.ptr();
}
```

Error:
```
In file included from /home/benchmark/src/pybind11/include/pybind11/pybind11.h:43:
In file included from /home/benchmark/src/pybind11/include/pybind11/attr.h:13:
/home/benchmark/src/pybind11/include/pybind11/cast.h:1858:38: error: non-const lvalue reference to type 'char' cannot bind to a temporary of type 'typename make_caster<char &>::cast_op_type<typename std::add_rvalue_reference<char &>::type>' (aka 'char')
        return std::forward<Func>(f)(cast_op<Args>(std::move(std::get<Is>(argcasters)))...);
                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/benchmark/src/pybind11/include/pybind11/cast.h:1840:35: note: in instantiation of function template specialization 'pybind11::detail::argument_loader<char &>::call_impl<void, void (*&)(char &), 0, pybind11::detail::void_type>' requested here
        std::move(*this).template call_impl<Return>(std::forward<Func>(f), indices{}, Guard{});
                                  ^
/home/benchmark/src/pybind11/include/pybind11/pybind11.h:154:52: note: in instantiation of function template specialization 'pybind11::detail::argument_loader<char &>::call<void, pybind11::detail::void_type, void (*&)(char &)>' requested here
                std::move(args_converter).template call<Return, Guard>(cap->f), policy, call.parent);
                                                   ^
/home/benchmark/src/pybind11/include/pybind11/pybind11.h:58:9: note: in instantiation of function template specialization 'pybind11::cpp_function::initialize<void (*&)(char &), void, char &, pybind11::name, pybind11::scope, pybind11::sibling>' requested here
        initialize(f, f, extra...);
        ^
/home/benchmark/src/pybind11/include/pybind11/pybind11.h:806:22: note: in instantiation of function template specialization 'pybind11::cpp_function::cpp_function<void, char &, pybind11::name, pybind11::scope, pybind11::sibling>' requested here
        cpp_function func(std::forward<Func>(f), name(name_), scope(*this),
                     ^
ret-reference.cpp:14:4: note: in instantiation of function template specialization 'pybind11::module::def<void (*)(char &)>' requested here
        m.def("foo_char",      &foo_char);
          ^
1 warning and 1 error generated.
```

--- a/include/pybind11/cast.h
+++ b/include/pybind11/cast.h
@@ -1216,6 +1216,7 @@ template <typename CharT> struct type_caster<CharT, enable_if_t<is_std_char_type
     using StringCaster = type_caster<StringType>;
     StringCaster str_caster;
     bool none = false;
+    CharT one_char = 0;
 public:
     bool load(handle src, bool convert) {
         if (!src) return false;
@@ -1243,7 +1244,7 @@ public:
     }
 
     operator CharT*() { return none ? nullptr : const_cast<CharT *>(static_cast<StringType &>(str_caster).c_str()); }
-    operator CharT() {
+    operator CharT&() {
         if (none)
             throw value_error("Cannot convert None to a character");
 
@@ -1267,7 +1268,8 @@ public:
             if (char0_bytes == str_len) {
                 // If we have a 128-255 value, we can decode it into a single char:
                 if (char0_bytes == 2 && (v0 & 0xFC) == 0xC0) { // 0x110000xx 0x10xxxxxx
-                    return static_cast<CharT>(((v0 & 3) << 6) + (static_cast<unsigned char>(value[1]) & 0x3F));
+                    one_char = static_cast<CharT>(((v0 & 3) << 6) + (static_cast<unsigned char>(value[1]) & 0x3F));
+                    return one_char;
                 }
                 // Otherwise we have a single character, but it's > U+00FF
                 throw value_error("Character code point not in range(0x100)");
@@ -1278,19 +1280,20 @@ public:
         // surrogate pair with total length 2 instantly indicates a range error (but not a "your
         // string was too long" error).
         else if (StringCaster::UTF_N == 16 && str_len == 2) {
-            char16_t v0 = static_cast<char16_t>(value[0]);
-            if (v0 >= 0xD800 && v0 < 0xE000)
+            one_char = static_cast<CharT>(value[0]);
+            if (one_char >= 0xD800 && one_char < 0xE000)
                 throw value_error("Character code point not in range(0x10000)");
         }
 
         if (str_len != 1)
             throw value_error("Expected a character, but multi-character string found");
 
-        return value[0];
+        one_char = value[0];
+        return one_char;
     }
 
     static constexpr auto name = _(PYBIND11_STRING_NAME);
-    template <typename _T> using cast_op_type = remove_reference_t<pybind11::detail::cast_op_type<_T>>;
+    template <typename _T> using cast_op_type = pybind11::detail::cast_op_type<_T>;
 };
 
 // Base implementation for std::tuple and std::pair

--- a/tests/test_builtin_casters.cpp
+++ b/tests/test_builtin_casters.cpp
@@ -50,7 +50,9 @@ TEST_SUBMODULE(builtin_casters, m) {
     // test_single_char_arguments
     m.attr("wchar_size") = py::cast(sizeof(wchar_t));
     m.def("ord_char", [](char c) -> int { return static_cast<unsigned char>(c); });
+    m.def("ord_char_lv", [](char &c) -> int { return static_cast<unsigned char>(c); });
     m.def("ord_char16", [](char16_t c) -> uint16_t { return c; });
+    m.def("ord_char16_lv", [](char16_t &c) -> uint16_t { return c; });
     m.def("ord_char32", [](char32_t c) -> uint32_t { return c; });
     m.def("ord_wchar", [](wchar_t c) -> int { return c; });
 

--- a/tests/test_builtin_casters.py
+++ b/tests/test_builtin_casters.py
@@ -44,6 +44,7 @@ def test_single_char_arguments():
     toolong_message = "Expected a character, but multi-character string found"
 
     assert m.ord_char(u'a') == 0x61  # simple ASCII
+    assert m.ord_char_lv(u'b') == 0x62
     assert m.ord_char(u'é') == 0xE9  # requires 2 bytes in utf-8, but can be stuffed in a char
     with pytest.raises(ValueError) as excinfo:
         assert m.ord_char(u'Ā') == 0x100  # requires 2 bytes, doesn't fit in a char
@@ -54,9 +55,11 @@ def test_single_char_arguments():
 
     assert m.ord_char16(u'a') == 0x61
     assert m.ord_char16(u'é') == 0xE9
+    assert m.ord_char16_lv(u'ê') == 0xEA
     assert m.ord_char16(u'Ā') == 0x100
     assert m.ord_char16(u'‽') == 0x203d
     assert m.ord_char16(u'♥') == 0x2665
+    assert m.ord_char16_lv(u'♡') == 0x2661
     with pytest.raises(ValueError) as excinfo:
         assert m.ord_char16(u'🎂') == 0x1F382  # requires surrogate pair
     assert str(excinfo.value) == toobig_message(0x10000)

